src/garage/tf/algos/npo.py

```diff
@@ -156,13 +170,100 @@ def init_opt(self):
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        last_return = None
+
+        for _ in runner.step_epochs():
+            runner.step_path = runner.obtain_samples(runner.step_itr)
+            last_return = self.train_once(runner.step_itr, runner.step_path)
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization given one batch of samples.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+
+        Returns:
+            numpy.float64: Average return.
+
+        """
+        # -- Stage: Calculate baseline
+        paths = [
+            dict(
+                observations=self._env_spec.observation_space.flatten_n(
+                    path['observations'])
+                if self._flatten_input else path['observations'],
+                actions=(
+                    self._env_spec.action_space.flatten_n(  # noqa: E126
+                        path['actions'])),
+                rewards=path['rewards'],
+                env_infos=path['env_infos'],
+                agent_infos=path['agent_infos'],
+                dones=path['dones']) for path in paths
+        ]
+
+        if hasattr(self._baseline, 'predict_n'):
+            baseline_predictions = self._baseline.predict_n(paths)
+        else:
+            baseline_predictions = [
+                self._baseline.predict(path) for path in paths
+            ]
+
+        # -- Stage: Pre-process samples based on collected paths
+        samples_data = paths_to_tensors(paths, self.max_path_length,
+                                        baseline_predictions, self._discount,
+                                        self._gae_lambda)
+
+        # -- Stage: Run and calculate performance of the algorithm
+        undiscounted_returns = log_performance(
+            itr,
+            TrajectoryBatch.from_trajectory_list(self._env_spec, paths),
+            discount=self._discount)
+        self._episode_reward_mean.extend(undiscounted_returns)
+        tabular.record('Extras/EpisodeRewardMean',
+                       np.mean(self._episode_reward_mean))
+
+        samples_data['average_return'] = np.mean(undiscounted_returns)
+
+        self.log_diagnostics(samples_data)
+        logger.log('Optimizing policy...')
+        self.optimize_policy(itr, samples_data)
+        return samples_data['average_return']
+
+    def log_diagnostics(self, paths):
+        """Log diagnostic information.
+
+        Args:
+            paths (list[dict]): A list of collected paths.
+
+        """
+        logger.log('Logging diagnostics...')
+        self.policy.log_diagnostics(paths)
+        self._baseline.log_diagnostics(paths)
+
+    # pylint: disable=unused-argument
```

Comment:
If a parameter is not used, prefer `del param` over pylint exception rule.

Reply:
@ryanjulian @naeioi 

I am working on a small PR to fix this. Found that the `itr` param of `optimize_policy()` is only used [here](https://github.com/rlworkgroup/garage/blob/master/src/garage/tf/algos/td3.py#L324). So simply removing `itr` from `optimize_policy()`will not work. 

In this context, should I just do `del itr`? Although `itr` is unused now, passing it to`optimize_policy()` can support potential future functionalities (e.g. logging) anyways.

What do you think?

Reply:
optimize_policy is not a public interface (i.e. these methods should actually be `_optimize_policy`, so it's free to be different between each algorithm).

Reply:
in general you should commit for the code you have, not the code you think you might have in the future.

i recommend removing it.

Reply:
`TD3` inherits from `DDPG`, and it seems unreasonable to have `TD3` overriding its parent's `optimize_policy()` function with a different signature. Both `optimize_policy()` are doing the same thing, but with different implementations and different inputs. So I also don't see how I can give them different names to avoid this inheritance conflict.

Is there a reason why `TD3` inherits from `DDPG`? If they don't have any special relationship (such as in an algorithm sense), I suggest:

1.  make `TD3` inherit from `OffPolicyRLAlgorithm` directly. 
2.  delete the abstract  `optimize_policy()` from `OffPolicyRLAlgorithm. As you suggested, this function should be private and sub-classes should not share the same interface.

Any thoughts?

Reply:
@maliesa96 and @krzentner  are working on removing OffPolicyRLAlgorithm entirely (for the exact same reasons we removed BatchPolopt), so I wouldn't spend too much time fretting over it.

If there a reason you can't remove itr from `NPO.optimize_policy()`? Is there an abstract `opimitze_policy` in NPO's inheritance tree?

Reply:
NPO doesn't have this issue. The conflict only exists in TD3 and DDPG inheritance tree.

So I guess I will make TD3 inherit from `OffPolicyRLAlgorithm`?

Reply:
That is probably harder than it looks (because of some bad design choices in OffPolicyRLAlgorithm, previously mentioned). 

What's the goal of your new PR? Eliminating `itr` from `optimizer_policy` everywhere?

Reply:
Ah I see. Yes the goal is to remove unused parameters from `optimize_policy()`. Most of them don't use `itr`, and [this one](https://github.com/rlworkgroup/garage/blob/master/src/garage/tf/algos/ddpg.py#L323) doesn't use neither `itr` nor `samples_data`.

This is not something urgent, as we can let `del itr` handle it and wait till `OffPolicyRLAlorithm` is refactored.

Reply:
Yeah, the story of `optimize_policy` is that someone tried to create an `RLAlgorithm` interface which included it, applied it (using `BatchPolopt` and `OffPolicyRLAlgorithm`), and then realized it doesn't really make sense for many algorithms.

The new `RLAlgorithm` interface only includes `train`. The motivation for this tiny interface, which forces each algorithm to do a lot, is that it's very difficult to define a single, universal control flow for an RL algorithm, and libraries which remove control flow from the user tend to make things difficult to understand.

Removing `BatchPolopt` and `OffPolicyRLAlgorithm` is basically unwinding this attempt. Going forward, anything which looks like `optimize_policy` can become a private function (`_optimize_policy`) inside the algorithms which were affected, only accessible. Whether `itr` appears in the signature is up to the algorithm, because it's a private function.

One of the biggest complaints from users about other RL libraries (why they don't adopt them and keep writing messy one-off code which often has bugs) is that they have to jump across 10 files just to understand what is happening in the algorithm. For the same reason, we're removing algorithm superclasses like BatchPolopt, so you can read the algorithm file and understand everything in an algorithm on "one pane of glass". The only reason we use inheritance on algorithms is for the most straightforward, shortest kind of inheritance, like when a single method gets overloaded in a simple way.

Perhaps a positive change would be to make everything which isn't `train` on all the algorithms underscore-private. This is also not urgent, but in retrospect the current situation tricks people into thinking there are public interfaces there when there are not.

Reply:
I am happy to put these in my PR to make functions except `train()` private. 

However, changing all `optimize_policy` to `_optimize_policy` won't solve the inheritance conflict in `TD3`. Because now `TD3` and `DDPG` both have `_optimize_policy`, but sub-class `TD3` has a different `_optimize_policy` signature.

So we either remove this inheritance relationship between `TD3` and `DDPG`, or give [class-specific](https://stackoverflow.com/questions/20261517/inheritance-of-private-and-protected-methods-in-python) names to `_optimize_policy` (e.g. `_TD3__optimize_policy`).

Please see my attempt to remove inheritance relationship: [#1516](https://github.com/rlworkgroup/garage/pull/1516)

Sorry this simple PR turns into this huge discussion. Really appreciate your help!!!